packages/dds/matrix/src/undoprovider.ts

```diff
@@ -164,11 +169,12 @@ export class MatrixUndoProvider<T> {
 		if (this.consumer !== undefined) {
 			this.consumer.pushToCurrentOperation({
 				revert: () => {
-					this.matrix.setCell(
-						this.rows.handleToPosition(rowHandle),
-						this.cols.handleToPosition(colHandle),
-						oldValue,
-					);
+					const row = this.rows.handleToPosition(rowHandle);
+					const col = this.cols.handleToPosition(colHandle);
+					// if the row/column no longer exists, we cannot set the cell
+					if (row < this.matrix.rowCount && col < this.matrix.colCount) {
```

Comment:
this bit doesn't look right to me and is probably the source of the unexpected test cases--handleToPosition needs to resolve certain "old" handles to undefined, which seems difficult to do with its current strategy. Just because they're in range doesn't mean they refer to a column/row that actually still exists. We actually have everything in place now to do the 3 year old suggestion of using merge-tree bookkeeping for undo-redo. It's not super efficient since we don't update all localSeq partials all the time, but it seems more reliable and the current method isn't efficient either.

[This commit](https://github.com/microsoft/FluidFramework/commit/8411e2c77eb4eaf375ea400acede2ccad49809c1) takes that approach and looks like it passes all tests with updated expectations. It still needs a little cleanup--I highlighted one issue with reconnect and undo/redo interleaving, and I think all of the localOpMetadata stuff can be simplified.

Reply:
if you want to work on this incrementally that's ok with me since the buggy behavior is probably better than crashing. Maybe it's worth adding a "redo" step to your added test cases though to confirm the user can get back to a reasonable state with this implementation

Reply:
I agree, and i'm not trying to fix that now. The commit you have looks like the right path, but there will still be cases where the position cannot be resolved, if it is removed remotely, and goes below the collab window, so the row/column is totally gone. I think that's what i see in https://dev.azure.com/fluidframework/internal/_workitems/edit/4202

For what i'm trying to do, i don't actually need this fixed, and my changes are only in the row column vectors, but the previous error made testing hard, as we hit them all the time with the stress modifications.

Reply:
makes sense